astcontext/blocks.go

```diff
@@ -0,0 +1,221 @@
+	inspect := func(n ast.Node) bool {
+		switch x := n.(type) {
+		case *ast.ForStmt:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.For)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.RangeStmt:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.For)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.IfStmt:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.If)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.SelectStmt:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.Select)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.SwitchStmt:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.Switch)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.FuncDecl:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.Type.Func)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.FuncLit:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.Type.Func)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.TypeSwitchStmt:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.Switch)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		}
+		return true
+	}
+
+	for _, file := range files {
+		// Inspect the AST and find all function declarements and literals
+		ast.Inspect(file, inspect)
+	}
+
+	return blocks
+}
+
+// BorderType represents Block's border type: left or right.
+type borderType string
+
+const (
+	leftBorderType  borderType = "left"
+	rightBorderType borderType = "right"
+)
+
+// Border is helper struct, that represents Block's border
+type border struct {
+	Block *Block
+	Pos   *Position
+	Type  borderType
+}
```

Comment:
I'm not sure what this is? Can you give an explicit example or a picture of what you mean?